zebrad/src/commands/start/sync.rs

```diff
@@ -35,86 +36,209 @@ where
+        // fanout parameter, to get resp1, ..., respF
         for _ in 0..self.fanout {
             let req = self.peer_set.ready_and().await.map_err(|e| eyre!(e))?.call(
                 zebra_network::Request::FindBlocks {
                     known_blocks: block_locator.clone(),
                     stop: None,
                 },
             );
-            self.tip_requests.push(req);
+            tip_futs.push(req);
         }
 
-        self.block_locator = block_locator;
-
-        Ok(())
-    }
-
-    /// Drive all chain extending futures to completion, request unknown blocks,
-    /// and extend prospective chain requests
-    pub async fn extend_tips(&mut self) -> Result<(), Report> {
-        let mut tip_set = HashSet::<BlockHeaderHash>::new();
-        while !self.tip_requests.is_empty() {
-            match self
-                .tip_requests
-                .next()
-                .await
-                .expect("expected: tip_requests is never empty")
-                .map_err::<Report, _>(|e| eyre!(e))
-            {
+        let mut download_set = HashSet::new();
+        while let Some(res) = tip_futs.next().await {
+            match res.map_err::<Report, _>(|e| eyre!(e)) {
                 Ok(zebra_network::Response::BlockHeaderHashes(hashes)) => {
                     info!(
                         new_hashes = hashes.len(),
                         in_flight = self.block_requests.len(),
                         downloaded = self.downloaded.len(),
                         "requested more hashes"
                     );
-                    let new_tip = hashes[0];
-                    let _ = tip_set.insert(new_tip);
-                    self.request_blocks(hashes).await?;
+
+                    if hashes.last() != Some(&super::GENESIS) {
+                        continue;
+                    }
+
+                    let mut hashes = hashes.into_iter().peekable();
+                    let new_tip = if let Some(tip) = hashes.next() {
+                        tip
+                    } else {
+                        continue;
+                    };
+
+                    // ObtainTips Step 3
+                    //
+                    // For each response, starting from the beginning of the
+                    // list, prune any block hashes already included in the
+                    // state, stopping at the first unknown hash to get resp1',
+                    // ..., respF'. (These lists may be empty).
+                    while let Some(&next) = hashes.peek() {
+                        let should_download = self
+                            .state
+                            .ready_and()
+                            .await
+                            .map_err(|e| eyre!(e))?
+                            .call(zebra_state::Request::Contains { hash: next })
+                            .await
+                            .is_err();
```

Comment:
This will ignore service errors, which I don't think is what we want, but we can fix it by changing the request to `Depth`.